lib/appium.js

```diff
@@ -57,110 +60,155 @@ class AppiumDriver extends BaseDriver {
     return this.sessions[sessionId];
   }
 
-  getDriverForCaps (caps) {
-    // TODO if this logic ever becomes complex, should probably factor out
-    // into its own file
-    if (!caps.platformName || !_.isString(caps.platformName)) {
-      throw new Error("You must include a platformName capability");
+  async getDriversMapping () {
+    if (this._driversMapping) {
+      return this._driversMapping;
     }
 
-    // we don't necessarily have an `automationName` capability,
-    if (caps.automationName) {
-      if (caps.automationName.toLowerCase() === this.validAutomations.SELENDROID.toLowerCase()) {
-        // but if we do and it is 'Selendroid', act on it
-        return SelendroidDriver;
-      } else if (caps.automationName.toLowerCase() === this.validAutomations.UIAUTOMATOR2.toLowerCase()) {
-        // but if we do and it is 'Uiautomator2', act on it
-        return AndroidUiautomator2Driver;
-      } else if (caps.automationName.toLowerCase() === this.validAutomations.XCUITEST.toLowerCase()) {
-        // but if we do and it is 'XCUITest', act on it
-        return XCUITestDriver;
-      } else if (caps.automationName.toLowerCase() === this.validAutomations.YOUIENGINE.toLowerCase()) {
-        // but if we do and it is 'YouiEngine', act on it
-        return YouiEngineDriver;
-      } else if (caps.automationName.toLowerCase() === this.validAutomations.ESPRESSO.toLowerCase()) {
-        log.warn('The Appium Espresso driver is currently in early beta and meant only for experimental usage. Its API is not yet complete or guaranteed to work. Please report bugs to the Appium team on GitHub.');
-        return EspressoDriver;
-      }
+    this._driversMapping = {
```

Comment:
All this can be a constant, at the top, rather than buried in here.

Reply:
we still need `this.validAutomations` to be available inside it, so I cannot move it out of the class. Perhaps, we could make it to a property

Reply:
Oh. Why in the world is that information set on the driver instance? That is pointless. It should just be exported as a constant from the base driver.

Reply:
I agree it is not very useful, but this is the fact we have now

Reply:
It seems to be there so that we can validate capabilities. But the normal drivers (outside AppiumDriver) will _never_ see an automation name that is invalid, since the code here on session start, _before_ the inner driver is picked, should fail if a bad one is sent.

Would you mind pulling that stuff (https://github.com/appium/appium-base-driver/blob/master/lib/basedriver/automation-names.js) over into here and just making a constant that this uses? I can make a PR to remove it from the base driver.

Reply:
I'd rather let you design it inside base driver.

Reply:
Base driver should know nothing of this. This is something specific to AppiumDriver. I'm happy to do it, though.

Reply:
In which file it would be better to define these constants for Appium driver? Right here, in appium.js or somewhere else?

Reply:
or feel free to alert the stuff to this PR